### ISSUE TITLE
man podman run: The documentation for user namespaces is incomprehensible

### ISSUE THREAD

**Issue by najamelan** [containers/libpod#1374]
```
--uidmap=map

       UID map for the user namespace.  Using this flag will run the container with user namespace enabled.  
       It conflicts with the --userns and --subuidname flags.

       The following example maps uids 0-2000 in the container to the uids 30000-31999 on the host and 
       gids 0-2000 in the container to the gids 30000-31999 on the host.

--gidmap=map

       GID map for the user namespace.  Using this flag will run the container with user namespace enabled.  
       It conflicts with the --userns and --subgidname flags.

       The following example maps uids 0-2000 in the container to the uids 30000-31999 on the host and 
       gids 0-2000 in the container to the gids 30000-31999 on the host.

--subgidname=name

       Name for GID map from the /etc/subgid file.  Using this flag will run the container with user namespace enabled.  
       This flag conflicts with --userns and --gidmap.

--subuidname=name 

       Name for UID map from the /etc/subuid file.  Using this flag will run the container with user namespace enabled. 
       This flag conflicts with --userns and --uidmap.

--userns=""

       Set the usernamespace mode for the container. The use of userns is disabled by default.

       host: use the host usernamespace and enable all privileged options (e.g., pid=host or --privileged). 
       ns: specify the usernamespace to use.
```
In general, UID and GID issues seem problematic in the container community, not only for podman. At least we can say they're confusing for newcomers. Probably the documentation (for podman run) could have a section explaining things a bit from a general perspective.

That being said, that's why I started looking into user namespaces and wondering whether podman supports them. The three sections above seem to be the documentation about user namespaces. It is not at all clear to me:
- why there's 3 different options
- what are the differences between them
- which one would I use in which situation
- their syntax (gidmap mentions an example, but there is none and userns has "", two quotes, what is that supposed to mean?)

Tracking issue: #1321

**Comment by vrothberg** [comment]
@giuseppe PTAL

**Comment by giuseppe** [comment]
they serve different purposes:

`--gidmap` is used to create a new usernamespace and specify the user namespace mapping to use, basically it is what you see in `/proc/self/gid_map`

`--subgidname` tells podman to read the gid mapping to use from the `/etc/subgid` file.   So for example if you have `--subgidname=foo`, podman will read any `foo:` in `/etc/subgid` and create a new usernamespace using this information for the mapping.

`--userns` tweaks how to create the usernamespace, for example you can say to use the host namespace, `--userns=host`, or to join an existing one: `--userns=ns:/proc/1337/ns/user`

**Comment by najamelan** [comment]
Ok I found some more options, added them above. So if I get it right, 

- gidmap, uidmap create a new namespace with the supplied map
- subuid, subguid create a new namespace with the map from /etc/subuid
- userns=host -> don't use a user namespace? Is this the same as running without specifying any of these options?
- userns=mynamespace use an existing namespace

I would propose the following:
1. That the documentation of `--gidmap` be changed to:
    ```
    --gidmap=container_gid:host_gid:amount <- I don't know if this is the actual syntax, since I can't figure it out from the man page, just guessing.
    --gidmap=0:30000:2000 

    Create a new user namespace with the supplied mapping. This option conflicts with the --userns and --subgidname flags. 
    If calling podman run as an unprivileged user, the user needs to have the right to use the mapping. See `man subuid`.
    The example maps gids 0-2000 in the container to the gids 30000-31999 on the host.
    ```

For `--subgidname` I would also mention that is creates the namespace. Also, for both those options, it might not yet be very important now, but in the future rootless containers will probably be more popular, and if I'm not mistaking the calling user needs to have the right to use a certain mapping (by having an entry in suid). I have added a line based on that assumption.

Right now podman doesn't work as non privileged user on my system... don't know why.

If `--userns=host` is the default behavior, without namespaces, I would formulate it as:
```
Use the namespace of the calling user on the host. This is the default behavior if no namespaces are enabled with other options. 
The processes in the container will have the same privileges on the host as the calling user.
```

Does this all make sense? Do the proposed changes seem clearer to you as well?

**Comment by rhatdan** [comment]
@najamelan Could you open a PR with your suggested changes.   I agree with you that this is confusing, and having this information from a user not familiar with the internals is helpful.

**Comment by giuseppe** [comment]
@najamelan yes thanks that looks much better.  One thing to clarify: `-gidmap=container_gid:host_gid:amount` can be specified multiple times so that different mappings are added to the container.

**Comment by vrothberg** [comment]
Closing as this issue has been fixed via #1380.
